Sources/Basic/Path.swift

```diff
@@ -0,0 +1,473 @@
+                //         up-front whether a path is absolute or relative.
+                let name = String(chars[nStart ..< nEnd])
+                let homeDir = gethomedir(user: name).characters
+                chars.replaceSubrange(chars.startIndex ..< nEnd, with: homeDir)
+            }
+        }
+        
+        // At this point we expect to have a path separator as first character.
+        assert(chars.first == PathSeparatorCharacter)
+
+        // Split the character array into parts, folding components as we go.
+        // As we do so, we count the number of characters we'll end up with in
+        // the normalized string representation.
+        var parts: [String.CharacterView] = []
+        var capacity = 0
+        for part in chars.split(separator: PathSeparatorCharacter) {
+            switch part.count {
+              case 0:
+                // Ignore empty path components.
+                continue
+              case 1 where part.first == ".":
+                // Ignore `.` path components.
+                continue
+              case 2 where part.first == "." && part.last == ".":
+                // If there's a previous part, drop it; otherwise, do nothing.
+                if let prev = parts.last {
+                    parts.removeLast()
+                    capacity -= prev.count
+                }
+              default:
+                // Any other component gets appended.
+                parts.append(part)
+                capacity += part.count
+            }
+        }
+        capacity += max(parts.count, 1)
+        
+        // Create an output buffer using the capacity we've calculated.
+        // FIXME: Determine whether this is the most efficient way to do it.
+        var result = ""
+        result.reserveCapacity(capacity)
+        
+        // Put the normalized parts back together again.
+        var iter = parts.makeIterator()
+        result.append(PathSeparatorCharacter)
+        if let first = iter.next() {
+            result.append(contentsOf: first)
+            while let next = iter.next() {
+                result.append(PathSeparatorCharacter)
+                result.append(contentsOf: next)
+            }
+        }
+        
+        // Sanity-check the result (including the capacity we reserved).
+        assert(!result.isEmpty, "unexpected empty string")
+        assert(result.characters.count == capacity, "count: " +
+            "\(result.characters.count), cap: \(capacity)")
+        
+        // Use the result as our stored string.
+        _impl = PathImpl(string: String(result))
+    }
+    
+    /// Directory component.  An absolute path always has a non-empty directory
+    /// component (the directory component of the root path is the root itself).
+    public var dirname: String {
+        return _impl.dirname!
+    }
+    
+    /// Last path component (including the suffix, if any).  it is never empty.
```

Comment:
Oh, I see -- I hadn't read the implementation yet. I personally would expect the base name to never contain "/" but I guess that was just my assumption. I don't have a strong opinion on the right answer here, I see that the POSIX utility says it is "/" while most APIs I have used say "".